File: backend/app/routers/stats.py
from fastapi import APIRouter, Depends

from app.core.deps import require_role
from app.db import mongo
from app.models.content import AdminStats, PublicStats

router = APIRouter(prefix="/api/v1/stats", tags=["stats"])


@router.get("/public", response_model=PublicStats)
async def public_stats():
    cursor = mongo.teams.find({})
    teams = await cursor.to_list(length=None)
    ideas = sum(1 for t in teams if t.get("level1", {}).get("submission_url"))
    ps_count = await mongo.problem_statements.count_documents({})
    return PublicStats(
        teams_registered=len(teams),
        ideas_submitted=ideas,
        problem_statements=ps_count,
        days_to_deadline=12,
    )


@router.get("/admin", response_model=AdminStats)
async def admin_stats(_coordinator: dict = Depends(require_role("coordinator"))):
    cursor = mongo.teams.find({})
    teams = await cursor.to_list(length=None)
    stages = {"registered": 0, "level1": 0, "level2": 0, "selected": 0}
    for t in teams:
        st = t.get("status", "registered")
        if st == "registered":
            stages["registered"] += 1
        elif st in ("l1_submitted", "l1_under_review", "l1_cleared", "l1_rejected"):
            stages["level1"] += 1
        elif st in ("l2_submitted", "l2_under_review", "l2_rejected"):
            stages["level2"] += 1
        elif st == "selected":
            stages["selected"] += 1

    return AdminStats(
        total_teams=len(teams),
        total_students=sum(len(t.get("members", [])) + 1 for t in teams),
        by_stage=stages,
        selected=stages["selected"],
    )

File: frontend/src/backend/stats.ts
import { Hono } from 'hono';
import type { AppEnv } from './types';
import { authMiddleware } from './auth';

export const statsRouter = new Hono<AppEnv>();

statsRouter.get('/public', async (c) => {
  const teamsCount = await c.env.DB.prepare('SELECT COUNT(*) as count FROM teams').first();
  const psCount = await c.env.DB.prepare('SELECT COUNT(*) as count FROM problem_statements').first();
  
  return c.json({
    teams_registered: (teamsCount?.count as number) || 0,
    ideas_submitted: (teamsCount?.count as number) || 0,
    problem_statements: (psCount?.count as number) || 0,
    days_to_deadline: 15
  });
});

statsRouter.get('/admin', authMiddleware, async (c) => {
  const user = c.get('user');
  if (user.role !== 'admin' && user.role !== 'spoc' && user.role !== 'coordinator') {
    return c.json({ detail: 'Forbidden' }, 403);
  }

  const teamsCount = await c.env.DB.prepare('SELECT COUNT(*) as count FROM teams').first();
  const studentsCount = await c.env.DB.prepare('SELECT COUNT(*) as count FROM team_members').first();
  
  const level1Count = await c.env.DB.prepare(
    "SELECT COUNT(*) as count FROM teams WHERE status IN ('l1_submitted', 'l1_cleared', 'l1_rejected', 'l2_submitted', 'selected', 'l2_rejected')"
  ).first();

  const level2Count = await c.env.DB.prepare(
    "SELECT COUNT(*) as count FROM teams WHERE status IN ('l2_submitted', 'selected', 'l2_rejected')"
  ).first();

  const selectedCount = await c.env.DB.prepare(
    "SELECT COUNT(*) as count FROM teams WHERE status = 'selected'"
  ).first();

  const totalTeams = (teamsCount?.count as number) || 0;
  const totalStudents = (studentsCount?.count as number) || 0;

  return c.json({
    total_teams: totalTeams,
    total_students: totalStudents,
    by_stage: {
      registered: totalTeams,
      level1: (level1Count?.count as number) || 0,
      level2: (level2Count?.count as number) || 0,
      selected: (selectedCount?.count as number) || 0
    },
    selected: (selectedCount?.count as number) || 0
  });
});
